test/distributed/pipelining/test_schedule_multiproc.py

```diff
@@ -60,38 +60,197 @@ def backend_str(cls) -> str:
+        ref_mod = copy.deepcopy(full_mod)
 
         x = torch.randn(batch_size, d_hid, device=self.device)
-        x_clone = x.clone()
+        with torch.no_grad():
+            y = ref_mod(x)
+            target = y + torch.randn(batch_size, d_hid, device=self.device)
 
-        num_microbatches = 2 * self.world_size
-        x_mb = x.chunk(num_microbatches)[0]
+        loss_fn = torch.nn.MSELoss(reduction="sum")
+        return full_mod, ref_mod, x, target, loss_fn
 
-        # Create a pipeline
-        split_spec = mod.split_spec if hasattr(mod, "split_spec") else None
-        pipe = pipeline(
-            mod,
-            mb_args=(x_mb,),
-            split_spec=split_spec,
-        )
+    def _create_single_stage_pipeline(self, mod, x, chunks, use_tracer=True):
+        """Create a single-stage pipeline using either tracer or manual stage creation."""
+        if use_tracer:
+            x_mb = x.chunk(chunks)[0]
+            split_spec = mod.split_spec if hasattr(mod, "split_spec") else None
+            pipe = pipeline(mod, mb_args=(x_mb,), split_spec=split_spec)
+            stage = pipe.build_stage(self.rank, self.device)
+            stage_module = pipe.get_stage_module(self.rank)
+            return stage, stage_module, [stage_module]
+        else:
+            # Manual stage creation
+            submod_name = f"layers.{self.rank}"
+            stage_module = mod.get_submodule(submod_name)
+            stage = PipelineStage(stage_module, self.rank, self.world_size, self.device)
+            return stage, stage_module, [stage_module]
+
+    def _create_multi_stage_pipeline(
+        self, mod, stages_per_rank, n_stages, stage_indices=None
+    ):
+        """Create multiple pipeline stages for interleaved schedules."""
+        if stage_indices is None:
+            stage_indices = [
+                self.rank + i * self.world_size for i in range(stages_per_rank)
+            ]
 
-        stage = pipe.build_stage(
-            self.rank,
-            self.device,
+        submod_names = [f"layers.{i}" for i in stage_indices]
+        stage_modules = [mod.get_submodule(submod_name) for submod_name in submod_names]
+        stages = [
+            PipelineStage(stage_module, stage_idx, n_stages, self.device)
+            for stage_module, stage_idx in zip(stage_modules, stage_indices)
+        ]
+        return stages, stage_modules, submod_names
+
+    def _run_reference_model(
+        self, ref_mod, x, target, loss_fn, num_iterations=2, **kwargs
+    ):
+        """Run reference model for specified iterations and return final output and loss."""
+        ref_out = None
+        ref_loss = None
+
+        for _ in range(num_iterations):
+            ref_mod.zero_grad()
+            ref_out = ref_mod(x, **kwargs)
+            ref_loss = loss_fn(ref_out, target)
+            ref_loss.backward()
+
+        return ref_out, ref_loss
+
+    def _run_schedule_steps(self, schedule, x, target=None, num_iterations=2, **kwargs):
```

Comment:
i'm a little dubious of the value of extracting the runner loop as a helper.  The other helpers all seem great to me.  is it more clear if the test just loops over calling the model and the ref?  I don't have a strong objection here.